### PR TITLE
add support for local metadata (option 2)

### PR DESCRIPTION
**Pull request by tomerd**

### CODE REVIEW

File: Sources/Logging/Logging.swift [hunk 1]
```diff
@@ -34,35 +34,39 @@ public struct Logger {
...
 
     @inlinable
-    func log(level: Logging.Level, message: @autoclosure () -> String, error: Error? = nil, file: StaticString = #file, function: StaticString = #function, line: UInt = #line) {
+    func log(level: Logging.Level, message: @autoclosure () -> String, metadata meta: @autoclosure () -> Logging.Metadata? = nil, error: Error? = nil, file: StaticString = #file, function: StaticString = #function, line: UInt = #line) {
         if self.logLevel <= level {
-            self.handler.log(level: level, message: message(), error: error, file: file, function: function, line: line)
+            var metadata = self.metadata
+            if let m = meta() {
+                metadata.merge(m, uniquingKeysWith: { _, new in new })
+            }
```
Comment (tomerd) [inline Sources/Logging/Logging.swift:42]: in this option we do the merge at the API level ^^

File: Sources/Logging/Logging.swift [hunk 2]
```diff
@@ -240,6 +244,11 @@ internal final class StdoutLogger: LogHandler {
 
     public init(label: String) {}
 
+    public func log(level: Logging.Level, message: String, metadata: Logging.Metadata, error: Error?, file: StaticString, function: StaticString, line: UInt) {
+        let prettyMetadata = !metadata.isEmpty ? self.prettify(metadata) : nil
```
Comment (tomerd) [inline Sources/Logging/Logging.swift:248]: the cost is the implementations will be forced to compute the metadata representation on every call, iow it moves to the hot path

**Comment by ktoso** [comment]
I'm strongly in favor of option 1, replied in more detail there: https://github.com/weissi/swift-server-logging-api-proposal/pull/16#issuecomment-452549526

This `option 2` conflates things somewhat and takes things out of the hands of logging impls which ultimately should be the ones in charge what happens to any data incoming to them. Also the lack of optimization potential in option 2 is not so nice.

👎 for this one, and 👍  for https://github.com/weissi/swift-server-logging-api-proposal/pull/16

**Comment by weissi** [comment]
@ktoso & @tomerd I agree with Konrad. `LogHandler`s might also not necessarily want to store the metadata as a dictionary internally so I think we shouldn't pre-merge for them. I'm going to merge #16

**Pull request closed without merging**